src/RuleInterface.php

```diff
@@ -14,9 +14,15 @@ interface RuleInterface
      *
      * @param string $userId The user ID. This should be a string representing the unique identifier of a user.
      * @param Item $item The role or permission that this rule is associated with.
+     * @param RuleFactoryInterface $ruleFactory Rule factory.
      * @param array $parameters Parameters passed to {@see CheckAccessInterface::userHasPermission()}.
      *
      * @return bool Whether the rule permits the auth item it is associated with.
      */
-    public function execute(string $userId, Item $item, array $parameters = []): bool;
+    public function execute(
+        string $userId,
+        Item $item,
+        RuleFactoryInterface $ruleFactory,
+        array $parameters = [],
```

Comment:
Suggest don't add new parameter. Replace `$parameters` to context object instead of.

```php
public function execute(string $userId, Item $item, Context $context): bool;

final class Context {
  public function getParams(): array;
  public function getParam(string $name): mixed;
  public function hasParam(string $name): bool;
  public function createRule(string $name): RuleInterface;
}
```

What do you think about it?